### PR TITLE
Tool recommendation

### PR DESCRIPTION
**Pull request by anuprulez**
This PR is inspired by https://github.com/galaxyproject/galaxy/pull/8714

tr

The old PR also shows tool recommendations in the workflow editor. But, the UI structure of the workflow's page has been updated (e.g. global client variable `workflow_globals` was used heavily to fetch the workflow, to find subtypes etc.). I am not sure how to align tool recommendation features to the new version of the workflow UI structure. I need to have communication between `Node.vue` (where a right-arrow icon in a tool's box can be specified) and `Index.vue` (where `manager` object is created) at `components/workflow/editor` to make it work again. Some help/advice here would be awesome. Thanks!

The current PR shows tool recommendations only after tool executions.

ping @bgruening

### CODE REVIEW

**Comment by jdavcs** [comment]
Just in case: I made this comment to the old PR right before it was closed, so this might be no longer relevant: https://github.com/galaxyproject/galaxy/pull/8714/files?file-filters%5B%5D=.yml#r384018957 If no longer relevant - ignore.

**Comment by anuprulez** [comment]
Fixed it. Thanks @ic4f

**Comment by anuprulez** [comment]
ping @guerler

File: lib/galaxy/dependencies/pipfiles/default/pinned-requirements.txt [hunk 1]
```diff
@@ -78,6 +78,7 @@ jmespath==0.9.4
 jsonpatch==1.25
 jsonpointer==2.0
 jsonschema==3.2.0
+keras==2.2.4
```
Comment (dannon) [inline lib/galaxy/dependencies/pipfiles/default/pinned-requirements.txt]: Can we swap this and tensorflow to conditional requirements based on `enable_tool_recommendations` in the config?

File: lib/galaxy/webapps/galaxy/api/workflows.py [hunk 2]
```diff
@@ -8,45 +8,35 @@
...
-    exceptions,
-    model,
-    util
-)
-from galaxy.managers import (
-    histories,
-    workflows
-)
+from galaxy import exceptions, model, util
+from galaxy.managers import histories, workflows
```
Comment (nsoranzo) [inline lib/galaxy/webapps/galaxy/api/workflows.py]: We prefer to have the `from ...` imports on separate lines to reduce merge conflicts. Can you revert these changes?

File: lib/galaxy/webapps/galaxy/api/workflows.py [hunk 3]
```diff
@@ -8,45 +8,35 @@
...
-from galaxy.webapps.base.controller import (
-    BaseAPIController,
-    SharableMixin,
-    url_for,
-    UsesStoredWorkflowMixin
-)
+from galaxy.web import (expose_api, expose_api_anonymous_and_sessionless,
+                        expose_api_raw, format_return_as_json)
+from galaxy.webapps.base.controller import (BaseAPIController, SharableMixin,
+                                            url_for, UsesStoredWorkflowMixin)
```
Comment (nsoranzo) [inline lib/galaxy/webapps/galaxy/api/workflows.py]: As above.

File: lib/galaxy/dependencies/pipfiles/default/pinned-requirements.txt (hunk 1 above)
Comment (anuprulez) [reply lib/galaxy/dependencies/pipfiles/default/pinned-requirements.txt]: Fixed it. Thanks a lot!

File: lib/galaxy/webapps/galaxy/api/workflows.py (hunk 2 above)
Comment (anuprulez) [reply lib/galaxy/webapps/galaxy/api/workflows.py]: Fixed them. Thanks a lot!

**Comment by anuprulez** [comment]
@dannon I made the Keras related dependencies conditional which now causes an API test case (https://jenkins.galaxyproject.org/job/docker-api-py3/2920/testReport/junit/lib.galaxy_test.api.test_workflows/WorkflowsApiTestCase/test_get_tool_predictions/) to fail because it tries to import Keras in `from keras.models import ...` https://github.com/galaxyproject/galaxy/blob/e3d028c186b3cf2159a2efaa7420da72a8444d32/lib/galaxy/webapps/galaxy/api/workflows.py#L679 which is not downloaded by default. I can write this import statement in a `try catch` block but it would make little sense for this test case. Any suggestions? 

I can also refactor the API test to have checks for different responses such as 200 and 400

Thank you!

**Comment by anuprulez** [comment]
I am not sure the failing test now is related to the PR

`galaxy_test.selenium.test_uploads.UploadsTestCase.test_upload_paired_list`

https://jenkins.galaxyproject.org/job/docker-selenium-py3/3219/

**Comment by jdavcs** [comment]
Rerunning it.

**Comment by anuprulez** [comment]
All tests passed!

File: client/galaxy/scripts/mvc/tool-recommendation.js [hunk 4]
```diff
@@ -0,0 +1,237 @@
+import Backbone from "backbone";
+import Utils from "utils/utils";
+import * as d3 from "d3";
+import { getAppRoot } from "onload/loadConfig";
+import $ from "jquery";
+
+var ToolRecommendationView = Backbone.View.extend({
```
Comment (guerler) [inline client/galaxy/scripts/mvc/tool-recommendation.js]: Sorry that I join so late on this topic but is it possible to integrate this as vue component?

File: client/galaxy/scripts/mvc/tool-recommendation.js [hunk 5]
```diff
@@ -0,0 +1,237 @@
...
+                .attr("class", "node")
+                .attr("transform", d => {
+                    return "translate(" + source.y0 + "," + source.x0 + ")";
+                })
+                .on("click", click);
+            nodeEnter
+                .append("circle")
+                .attr("r", 1e-6)
+                .style("fill", d => {
+                    return d._children ? "lightsteelblue" : "#fff";
```
Comment (guerler) [inline client/galaxy/scripts/mvc/tool-recommendation.js]: We probably should move these fixed styles and dimensions above, to the top of the file or ideally to css.

File: client/galaxy/scripts/mvc/tool-recommendation.js (hunk 4 above)
Comment (anuprulez) [reply client/galaxy/scripts/mvc/tool-recommendation.js]: I will look into it. Thanks!

**Comment by anuprulez** [comment]
I am not sure if the failing tests are related to the PR.
@guerler the client-side is in VueJS now

File: client/galaxy/scripts/components/ToolRecommendation.vue [hunk 6]
```diff
@@ -0,0 +1,282 @@
...
+                toolId = toolIdSlash[toolIdSlash.length - 2];
+            }
+            return toolId;
+        }
+    },
+    methods: {
+        loadRecommendations: function() {
+            const self = this;
+            const toolId = this.getToolId;
+            Utils.request({
```
Comment (guerler) [inline client/galaxy/scripts/components/ToolRecommendation.vue]: Please use axios here.

File: client/galaxy/scripts/components/ToolRecommendation.vue [hunk 7]
```diff
@@ -0,0 +1,282 @@
...
+}
+
+.tree-size {
+    width: 100%;
+    height: 450px;
+}
+
+.link {
+    fill: none;
+    stroke: #c2ebc2;
```
Comment (guerler) [inline client/galaxy/scripts/components/ToolRecommendation.vue]: Please remove fixed colors and pixel based styles if possible and consider to place all constants, and fixed styles into a separate css file for better maintenance and style control.

File: lib/galaxy/webapps/galaxy/api/workflows.py [hunk 8]
```diff
@@ -631,6 +638,235 @@ def build_module(self, trans, payload={}):
             'post_job_actions'  : module.get_post_job_actions(inputs)
         }
 
+    @expose_api
+    def get_tool_predictions(self, trans, payload, **kwd):
```
Comment (guerler) [inline lib/galaxy/webapps/galaxy/api/workflows.py:637]: I recommend to place the business logic for this feature into a separate module e.g. lib/galaxy/tools/recommendations.py and then import it into this api module, to improve maintainability and code separation.

File: client/galaxy/scripts/components/ToolRecommendation.vue [hunk 9]
```diff
@@ -0,0 +1,282 @@
...
+            You have used {{ getToolId }} tool. {{ deprecatedMessage }}
+        </div>
+    </div>
+</template>
+
+<script>
+import Utils from "utils/utils";
+import * as d3 from "d3";
+import { getAppRoot } from "onload/loadConfig";
+import $ from "jquery";
```
Comment (guerler) [inline client/galaxy/scripts/components/ToolRecommendation.vue]: Jquery should not be needed in this component, same for importing the Utils module above.

File: client/galaxy/scripts/components/ToolRecommendation.vue [hunk 10]
```diff
@@ -0,0 +1,282 @@
...
+                                }
+                            }
+                        }
+                        filteredData["o_extensions"] = predData["o_extensions"];
+                        filteredData["name"] = predData["name"];
+                        filteredData["children"] = filteredChildren;
+                        if (filteredChildren.length > 0 && this.deprecated === false) {
+                            self.renderD3Tree(filteredData);
+                        } else if (this.deprecated === true) {
+                            this.deprecatedMessage = predData["message"];
```
Comment (guerler) [inline client/galaxy/scripts/components/ToolRecommendation.vue]: Use `prepData.message` instead.

File: client/galaxy/scripts/components/ToolRecommendation.vue [hunk 11]
```diff
@@ -0,0 +1,282 @@
...
+                        }
+                    }
+                }
+            });
+        },
+        renderD3Tree: function(predictedTools) {
+            const margin = { top: 20, right: 30, bottom: 20, left: 250 },
```
Comment (guerler) [inline client/galaxy/scripts/components/ToolRecommendation.vue]: I am a little worried about this margins, if these constants are needed please specifiy them explicitly at the top of the file as constants.

**Comment by anuprulez** [comment]
ping @guerler

File: client/galaxy/style/scss/base.scss [hunk 12]
```diff
@@ -1650,3 +1650,36 @@ body.reports {
     bottom: 0;
     background: url(../../images/largespinner.gif) no-repeat center center fixed;
 }
+
+/* Used for tree in tool recommendations */
+.node {
```
Comment (guerler) [inline client/galaxy/style/scss/base.scss]: This css should be placed into the `workflow.scss` file.

**Comment by guerler** [comment]
Can you describe, how to test this, after setting `enable_tool_recommendations` to true?

**Comment by anuprulez** [comment]
ping @guerler 

tr

File: client/galaxy/scripts/mvc/tool/tool-form.js [hunk 13]
```diff
@@ -267,6 +268,19 @@ const View = Backbone.View.extend({
                     }
                 }
                 this.$el.append(this._templateSuccess(response, job_def));
+                const enable_tool_recommendations = window.Galaxy.config.enable_tool_recommendations;
```
Comment (guerler) [inline client/galaxy/scripts/mvc/tool/tool-form.js]: Please use the `Galaxy` instance you get from `getGalaxyInstance()` instead of accessing the global object.

File: client/galaxy/scripts/components/ToolRecommendation.vue [hunk 14]
```diff
@@ -0,0 +1,243 @@
...
+            if (toolId.indexOf("/") > 0) {
+                const toolIdSlash = toolId.split("/");
+                toolId = toolIdSlash[toolIdSlash.length - 2];
+            }
+            return toolId;
+        }
+    },
+    methods: {
+        loadRecommendations: function() {
+            const self = this,
```
Comment (guerler) [inline client/galaxy/scripts/components/ToolRecommendation.vue]: `self` is not needed here. Please use `;` to separate the declarations.

File: client/galaxy/scripts/components/ToolRecommendation.vue [hunk 15]
```diff
@@ -0,0 +1,243 @@
...
+        return {
+            deprecated: null,
+            deprecatedMessage: ""
+        };
+    },
+    created: function() {
+        this.loadRecommendations();
+    },
+    computed: {
+        getToolId: function() {
```
Comment (guerler) [inline client/galaxy/scripts/components/ToolRecommendation.vue]: Please use es6 functions `() => {}`.

File: client/galaxy/scripts/components/ToolRecommendation.vue [hunk 16]
```diff
@@ -0,0 +1,243 @@
...
+                                        ) {
+                                            compatibleTools[nameObj[1]["tool_id"]] = nameObj[1]["name"];
+                                            break;
+                                        }
+                                    }
+                                }
+                            }
+                            for (const id in compatibleTools) {
+                                for (const nameObj of children.entries()) {
+                                    if (nameObj[1]["tool_id"] === id) {
```
Comment (guerler) [inline client/galaxy/scripts/components/ToolRecommendation.vue]: Please use `nameObj[1].tool_id` instead.

File: client/galaxy/scripts/components/ToolRecommendation.vue [hunk 17]
```diff
@@ -0,0 +1,244 @@
...
+                                self.deprecatedMessage = predData.message;
+                            }
+                        }
+                    });
+                });
+        },
+        renderD3Tree: function(predictedTools) {
+            const duration = 750,
+                x = 620,
+                y = 260,
```
Comment (guerler) [inline client/galaxy/scripts/components/ToolRecommendation.vue]: Please use individual complete lines to declare constants like e.g. `const x = 620;`

File: client/galaxy/scripts/components/ToolRecommendation.vue (hunk 15 above)
Comment (anuprulez) [reply client/galaxy/scripts/components/ToolRecommendation.vue]: I tried using this format of specifying functions in `computed` and `methods` but the object `this` stopped getting recognised. Other places the ES6 format works

File: client/galaxy/style/scss/base.scss [hunk 18]
```diff
@@ -1650,3 +1650,36 @@ body.reports {
     bottom: 0;
     background: url(../../images/largespinner.gif) no-repeat center center fixed;
 }
+
+/* Used for tree in tool recommendations */
```
Comment (guerler) [inline client/galaxy/style/scss/base.scss:1654]: Please wrap all these classes into something like ".tool-recommendation-view { }".

**Comment by guerler** [comment]
I get the following error when running a tool multiple times. Once this error is triggered the entire server seems to freeze. 

`Traceback (most recent call last):
  File "lib/galaxy/web/framework/decorators.py", line 295, in decorator
    rval = func(self, trans, *args, **kwargs)
  File "lib/galaxy/webapps/galaxy/api/workflows.py", line 657, in get_tool_predictions
    is_model_set = self.tool_recommendations.set_model(trans, remote_model_url)
  File "lib/galaxy/tools/tool_recommendations.py", line 72, in set_model
    self.loaded_model.set_weights(model_weights)
  File "/Users/guerler/galaxy/.venv/lib/python2.7/site-packages/keras/engine/network.py", line 508, in set_weights
    K.batch_set_value(tuples)
  File "/Users/guerler/galaxy/.venv/lib/python2.7/site-packages/keras/backend/tensorflow_backend.py", line 2470, in batch_set_value
    get_session().run(assign_ops, feed_dict=feed_dict)
  File "/Users/guerler/galaxy/.venv/lib/python2.7/site-packages/tensorflow_core/python/client/session.py", line 956, in run
    run_metadata_ptr)
  File "/Users/guerler/galaxy/.venv/lib/python2.7/site-packages/tensorflow_core/python/client/session.py", line 1123, in _run
    e.args[0])
TypeError: Cannot interpret feed_dict key as Tensor: Tensor Tensor("Placeholder_4:0", shape=(114, 342), dtype=float32) is not an element of this graph.`

**Comment by anuprulez** [comment]
> I get the following error when running a tool multiple times. Once this error is triggered the entire server seems to freeze.
> 
> `Traceback (most recent call last): File "lib/galaxy/web/framework/decorators.py", line 295, in decorator rval = func(self, trans, *args, **kwargs) File "lib/galaxy/webapps/galaxy/api/workflows.py", line 657, in get_tool_predictions is_model_set = self.tool_recommendations.set_model(trans, remote_model_url) File "lib/galaxy/tools/tool_recommendations.py", line 72, in set_model self.loaded_model.set_weights(model_weights) File "/Users/guerler/galaxy/.venv/lib/python2.7/site-packages/keras/engine/network.py", line 508, in set_weights K.batch_set_value(tuples) File "/Users/guerler/galaxy/.venv/lib/python2.7/site-packages/keras/backend/tensorflow_backend.py", line 2470, in batch_set_value get_session().run(assign_ops, feed_dict=feed_dict) File "/Users/guerler/galaxy/.venv/lib/python2.7/site-packages/tensorflow_core/python/client/session.py", line 956, in run run_metadata_ptr) File "/Users/guerler/galaxy/.venv/lib/python2.7/site-packages/tensorflow_core/python/client/session.py", line 1123, in _run e.args[0]) TypeError: Cannot interpret feed_dict key as Tensor: Tensor Tensor("Placeholder_4:0", shape=(114, 342), dtype=float32) is not an element of this graph.`

I ran a tool 15 times back-to-back and I did not get this error. I am using python=3.6 for my local 
Galaxy instance. Can you try using python=3.6 if you get the same error? I see you are using python=2.7

Moreover, I opened two browser instances and sent back-to-back 10 jobs from both sessions, still 
I did not get this error.

File: lib/galaxy/dependencies/conditional-requirements.txt [hunk 19]
```diff
@@ -28,3 +28,7 @@ watchdog
 python-gitlab
 pygithub
 influxdb
+
+# Deep learning packages for tool recommendation
+keras==2.2.4
+tensorflow==1.15.0
```
Comment (bgruening) [inline lib/galaxy/dependencies/conditional-requirements.txt]: @anuprulez can we use the 1.15.2 version here?

Comment (anuprulez) [reply lib/galaxy/dependencies/conditional-requirements.txt]: @bgruening I can but it will not make much difference

File: doc/source/admin/galaxy_options.rst [hunk 20]
```diff
@@ -4042,4 +4042,67 @@
...
+
+:Description:
+    Set remote path of the trained model (HDF5 file) for tool
+    recommendation.
+:Default: ``https://github.com/galaxyproject/galaxy-test-data/raw/master/tool_recommendation_model.hdf5``
+:Type: str
+
+
+~~~~~~~~~~~~~~~~~~~~~~~~
+``topk_recommendations``
```
Comment (guerler) [inline doc/source/admin/galaxy_options.rst:4069]: Can we remove this and only display significant results?

Comment (bgruening) [reply doc/source/admin/galaxy_options.rst:4069]: It can happen that there are many significant results.

File: client/galaxy/scripts/components/ToolRecommendation.vue [hunk 21]
```diff
@@ -0,0 +1,243 @@
...
+                            if (filteredChildren.length > 0 && this.deprecated === false) {
+                                this.renderD3Tree(filteredData);
+                            } else if (this.deprecated === true) {
+                                this.deprecatedMessage = predData.message;
+                            }
+                        }
+                    });
+                });
+        },
+        renderD3Tree: predictedTools => {
```
Comment (guerler) [inline client/galaxy/scripts/components/ToolRecommendation.vue]: Please use `renderD3Tree(predictedTools) { ... }` instead.

File: doc/source/admin/galaxy_options.rst (hunk 20 above)
Comment (guerler) [reply doc/source/admin/galaxy_options.rst:4069]: Ok

File: client/galaxy/scripts/components/ToolRecommendation.vue [hunk 22]
```diff
@@ -0,0 +1,243 @@
+<template>
+    <div id="tool-recommendation" class="tool-recommendation-view">
+        <div v-if="deprecated == false" class="infomessagelarge">
```
Comment (guerler) [inline client/galaxy/scripts/components/ToolRecommendation.vue]: Please use "!deprecated" and "!!deprecated" or v-else instead.

**Comment by guerler** [comment]
Ok, I retested it with the version used on the EU server and did not encounter this error again. Please resolve the remaining minor comments. Please also rename the file to "recommendations.py" since it is already in the "tools" directory. I am also wondering if we really need all 5 config options for this feature but everything else is fine with me 👍. Thank you for contributing this cool feature.

**Comment by anuprulez** [comment]
@guerler The config options `admin_tool_recommendations_path` and `overwrite_model_recommendations` are useful when an admin wants to change the predictions made by the deep learning model. For example, when newly added tools need to be promoted, this feature would come in handy - to overwrite the predictions by hand-chosen tools. The other
option `topk_recommendations` is useful when there is a need to show less or more number of predictions which can be a useful option for admins. They do not need to change in the code/API. The model predicts a fixed number of possible tools (currently set to 20).

Other comments are fixed now. Thanks a lot for your review and comments!

**Review by guerler — APPROVED** [review]

File: client/galaxy/scripts/components/ToolRecommendation.vue [hunk 23]
```diff
@@ -0,0 +1,243 @@
...
+    <div id="tool-recommendation" class="tool-recommendation-view">
+        <div v-if="!deprecated" class="infomessagelarge">
+            <h4>Tool recommendation</h4>
+            You have used {{ getToolId }} tool. For further analysis, you could try using the following/recommended
+            tools. The recommended tools are shown in the decreasing order of their scores predicted using machine
+            learning analysis on workflows. A tool with a higher score (closer to 100%) may fit better as the following
+            tool than a tool with a lower score. Please click on one of the following/recommended tools to open its
+            definition.
+        </div>
+        <div v-else-if="deprecated" class="warningmessagelarge">
```
Comment (guerler) [inline client/galaxy/scripts/components/ToolRecommendation.vue]: This is equivalent to "v-else".

**Comment by guerler** [comment]
Requires executing `make client-format`.

**Comment by anuprulez** [comment]
The test which is failing now failed a few times before as well. It does not seem to be related to this PR. Can someone restart the build?

**Comment by anuprulez** [comment]
@guerler thanks!
